### ISSUE TITLE
Orleans test host failed to load Microsoft.SqlServer.Types

### ISSUE THREAD

**Issue by saitonakamura** [dotnet/orleans#2710]
When trying to add `Microsoft.SqlServer.Types` nuget package to Orleans host and therefore load and use it in grains it fails with this errors

```
Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
Exception = System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\msvcr100.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].


[2017-02-09 14:18:20.723 GMT     6      ERROR   101716  AssemblyLoader.Client   ]       !!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.
Exc level 0: System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)
[2017-02-09 14:18:20.756 GMT     6      INFO    100000  AssemblyLoader.Client   ]       User assembly ignored: c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\msvcr100.dll
        * An unexpected exception occurred while attempting to load an assembly.
Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
Exception = System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\SqlServerSpatial110.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].


[2017-02-09 14:18:20.757 GMT     6      ERROR   101716  AssemblyLoader.Client   ]       !!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.
Exc level 0: System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)
[2017-02-09 14:18:20.767 GMT     6      INFO    100000  AssemblyLoader.Client   ]       User assembly ignored: c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x64\SqlServerSpatial110.dll
        * An unexpected exception occurred while attempting to load an assembly.
Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
Exception = System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\msvcr100.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].


[2017-02-09 14:18:20.768 GMT     6      ERROR   101716  AssemblyLoader.Client   ]       !!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.
Exc level 0: System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\msvcr100.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)
[2017-02-09 14:18:20.775 GMT     6      INFO    100000  AssemblyLoader.Client   ]       User assembly ignored: c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\msvcr100.dll
        * An unexpected exception occurred while attempting to load an assembly.
Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
Exception = System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\SqlServerSpatial110.dll'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].


[2017-02-09 14:18:20.776 GMT     6      ERROR   101716  AssemblyLoader.Client   ]       !!!!!!!!!! An unexpected exception occurred while attempting to load an assembly.
Exc level 0: System.BadImageFormatException: Could not load file or assembly 'file:///c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\SqlServerSpatial110.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.ReflectionOnlyLoadFrom(String assemblyFile)
   at Orleans.Runtime.AssemblyLoader.ReflectionOnlyLoadAssembly(String pathName, Assembly& assembly, String[]& complaints)
[2017-02-09 14:18:20.786 GMT     6      INFO    100000  AssemblyLoader.Client   ]       User assembly ignored: c:\users\mikhail_bashurov\documents\visual studio 2015\Projects\TestOrleansSqlServerTypes\Host\bin\Debug\SqlServerTypes\x86\SqlServerSpatial110.dll
        * An unexpected exception occurred while attempting to load an assembly.
```

However it feels fine in simple console application

[REPRO is here](https://github.com/saitonakamura/orleans-sql-server-types-test), it's based on 

- [minimal Orleans application tutorial](https://dotnet.github.io/orleans/Tutorials/Minimal-Orleans-Application.html), 
- installing `Microsoft.SqlServer.Types` nuget package on the `Host`, `GrainCollection` and `GrainInterfaces` projects
- adding `SqlServerTypes.Utilities.LoadNativeAssemblies(AppDomain.CurrentDomain.BaseDirectory);` to `Program.cs`

I suggest that it has something to do with separate domain for host

**Comment by saitonakamura** [comment]
I've already tried to launch host in x64 and x86 since `Microsoft.SqlServer.Types` has different builds, but it doesn't help either

**Comment by attilah** [comment]
Please ensure that you've the VC++ 2015 redists installed for x86 and x64 too.

https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads

**Comment by attilah** [comment]
I took a second look at the stacktraces. Which version of Orleans are you using? Since we enhanced the loading some time ago to get rid of ```BadImageException```, by adding a process compatibility check during the loading, also we're still explicitly catching ```BadImageException```, ```FileLoadException``` during reflection only loading.

**Comment by attilah** [comment]
I've modified the check for native library loading, and fixed in #2714. Thanks for reporting, it was taken care of a totally different 'not yet PR' branch.

**Comment by saitonakamura** [comment]
How long will it take to get it on nuget (I suppose it's about next beta/non-beta release)?
I'm asking more about dimensions, is it weeks or months?
